src/array/cuda/spmm.cu

```diff
@@ -282,6 +282,107 @@ void CusparseCsrmm2(
   if (valptr)
     device->FreeWorkspace(ctx, valptr);
 }
+
+/*! Cusparse implementation of SpMM on Csr format. */
+template <typename DType, typename IdType>
+void CusparseCsrmm2Hetero(
+    const DLContext& ctx,
+    const CSRMatrix& csr,
+    const DType* B_data, const DType* A_data,
+    DType* C_data,
+    int x_length,
+    cudaStream_t strm_id) {
+  // We use csrmm2 to perform following operation:
+  // C = A x B, where A is a sparse matrix in csr format, B is the dense matrix for node
+  // feature tensor. However, since cusparse only supports column-major, while our tensor
+  // is stored in row-major, the actual computation is:
+  // C = trans(A x trans(B)).
+  // Currently, we use cublasXgeam to implement transposition and allocate intermediate
+  // workspace memory for this.
+  const int m = csr.num_rows;
+  const int n = x_length;
+  const int k = csr.num_cols;
+  const int nnz = csr.indices->shape[0];
+  const DType alpha = 1.0;
+  const DType beta = 0.0;
+  // device
+  auto device = runtime::DeviceAPI::Get(ctx);
+  auto* thr_entry = runtime::CUDAThreadEntry::ThreadLocal();
+  // allocate cusparse handle if needed
+  if (!thr_entry->cusparse_handle) {
+    CUSPARSE_CALL(cusparseCreate(&(thr_entry->cusparse_handle)));
+  }
+  CUSPARSE_CALL(cusparseSetStream(thr_entry->cusparse_handle, strm_id));
+  // all one data array
+  DType* valptr = nullptr;
+  if (!A_data) {
+    valptr = static_cast<DType*>(device->AllocWorkspace(ctx, nnz * sizeof(DType)));
```

Comment:
Using a workspace allocation on stream different than the one associated with that context (which currently is always stream 0), could potentially cause race conditions between streams, as the workspace will get freed before the GPU is done with it, and it may be given to a different stream while the other stream is still using it.

Reply:
In cuda 11.2 there is a stream based allocator which would solve this issue nicely, but for now I think you'll have to allocate all of the memory before creating these streams, and free it only after destroying the streams.

Reply:
Allocating memory for all the streams might cause OOM. Is there a way to keep it flexible? For example, if there's not enough resource, limit the number of streams and only allocate memory for those streams?

Reply:
Right now, I don't think there is. You would need to know whether the WorkspacePool has an allocation that can be used, or a new one would need to be made. Same thing with the TensorAdapter.

In general though, I think restricting parallelism based on the amount of available memory will be problematic, both in terms of code complexity and performance. I think OOM error, and forcing the user to choose a different batch size/configuration, is better than running slower. We could maybe make an environment variable to control the maximum number of stream, and thus let the user control the amount of parallelism. But this would still require stream-based allocations to limit the concurrent memory usage.